notebooks/working/3.0-mab-independent-variable-simulation.md

```diff
@@ -34,487 +47,269 @@ import random
+## FindLongDataDist
+## Function for checking length
+def isUnique(var_values):
+    return len(var_values.unique()) == 1
+
+## Function for checking if variable is categorical
+def isCategorical(var_type):
+    return var_type == 'categorical'
+
+## Function to get distribution for unique values:
+def getConstantDist(ind_var, var_val):
+    constant_dict = defaultdict(dict)
+    constant_dict[ind_var]['distribution'] = 'constant'
+    constant_dict[ind_var]['parameters'] = var_val.unique()
+    return constant_dict
+
+## Function to get distribution for categorical variables:
+def getCategoricalDist(ind_var, var_val):
+    categ_dict = defaultdict(dict)
+    categ_dict[ind_var]['distribution'] = 'categorical'
+    # Count frequency of values and store it as paramater of distribution
+    np_array_range = np.arange(var_val.max()+1)
+    array_bincount = np.bincount(var_val)
+    probs = array_bincount / len(var_val)
+    categ_dict[ind_var]['parameters'] = [np_array_range,
+                                      probs]
+    return categ_dict
+
+
+def getContinuousDist(ind_var, var_val, cont_dists):
+    cont_dict = defaultdict(dict)    
+    fitter_object = Fitter(data=var_val,
+                           distributions=cont_dists,
+                           timeout=30)
+    fitter_object.fit()
+    # Get the best distribution and store in dictionary
+    BestDict = fitter_object.get_best()
+    cont_dict[ind_var]['distribution'] = list(BestDict.items())[0][0]
+    cont_dict[ind_var]['parameters'] = list(BestDict.items())[0][1]
+    return cont_dict
+
+## Function for looping around individual specific variables
+def indSpecDist(data_long, obs_id_col, ind_spec, var_types):
+    ind_spec_dict = defaultdict(dict)
+    for ind_var in ind_spec:
+        # generate array of values for individual specific variable
+        var_val = pd.Series([(data_long.loc[data_long[obs_id_col] == x][ind_var].unique()[0]) for x in data_long[obs_id_col].unique()])
+        var_type = var_types[ind_var]
+        # Get distribution if variable is categorical
+        if isCategorical(var_type):
+            # If only one category
+            if isUnique(var_val):
+                ind_var_dict = getConstantDist(ind_var, var_val)
+            # If more than one category
+            else:
+                ind_var_dict = getCategoricalDist(ind_var, var_val)
+        else:
+            ind_var_dict = getContinuousDist(ind_var, var_val, cont_dists)
+        ind_spec_dict.update(ind_var_dict)
+    return ind_spec_dict
+
+## Functions to replace code for getting
+## distributions for alternative specific
+## variables
+
+def getConstantDistAlt(alt_var, alt_data, alt_name_dic, alt):
+    constant_dict = defaultdict(dict)
+    # Add name of alternative to variable and store distriburion & parameters
+    constant_dict[alt_var+'_'+alt_name_dic[alt]]['distribution'] = 'constant'
+    constant_dict[alt_var+'_'+alt_name_dic[alt]]['parameters'] = alt_data[alt_var].unique()
```

Comment:
To the best extent possible, try to not repeat yourself. For example,
```suggestion
    alt_specific_var_name = alt_var + '_' + alt_name_dic[alt]
    constant_dict[alt_specific_var_name]['distribution'] = 'constant'
    constant_dict[alt_specific_var_name]['parameters'] = alt_data[alt_var].unique()
```

Reply:
As a matter of fact, you end up repeating this same pattern across multiple functions (for example, in `getCategoricalDistAlt` and `getContinuousDistAlt`.

This suggests that you should separate this logic into its own function. One might do:
```suggestion
    alt_specific_var_name =\
        get_alt_specific_variable_name(alt_var, alt_name_dict, alt)
    constant_dict[alt_specific_var_name]['distribution'] = 'constant'
    constant_dict[alt_specific_var_name]['parameters'] = alt_data[alt_var].unique()
```
You can define the separated function as follows:
```
def get_alt_specific_variable_name(var_name, alt_name_map, alt_id):
    return var_name + '_' + alt_name_map[alt_id]
```

Reply:
Got it! Whenever a logic repeats, isolate it.